File: c++/5014.cpp
#include <iostream>
#include <queue>

using namespace std;

int F, S, G, U, D;

bool isVisited[1000001];

int bfs() {
  queue<pair<int, int>> q;
  q.push({S, 0});
  isVisited[S] = true;

  while (!q.empty()) {
    int num = q.front().first;
    int cnt = q.front().second;
    q.pop();

    if (num == G) return cnt;

    int next1 = num + U;
    if (1 <= next1 && next1 <= F && !isVisited[next1]) {
      isVisited[next1] = true;
      q.push({next1, cnt + 1});
    }

    int next2 = num - D;
    if (1 <= next2 && next2 <= F && !isVisited[next2]) {
      isVisited[next2] = true;
      q.push({next2, cnt + 1});
    }
  }

  return -1;
}

int main() {
  freopen("example.txt", "r", stdin);

  cin >> F >> S >> G >> U >> D;

  int answer = bfs();
  if (answer == -1)
    cout << "use the stairs";
  else
    cout << answer;

  return 0;
}

File: js/5014.js
const fs = require('fs');

let [F, S, G, U, D] = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split(' ')
  .map(Number);

const isVisited = Array(1000001).fill(false);

console.log(bfs());

function bfs() {
  const q = [];
  q.push([S, 0]);
  isVisited[S] = true;

  while (q.length !== 0) {
    const [num, cnt] = q.shift();
    if (num === G) {
      return cnt;
    }

    for (let e of [num + U, num - D]) {
      if (1 <= e && e <= F && !isVisited[e]) {
        isVisited[e] = true;
        q.push([e, cnt + 1]);
      }
    }
  }

  return 'use the stairs';
}

// while (S !== G) {
//   if (S < G) {
//     S += U;
//   } else if (S > G) {
//     S -= D;
//   }
//   cnt++;
// }

// console.log(cnt);
